### PR TITLE
Add proposed field names to the mapped named jar

### PR DESCRIPTION
**Pull request by liach**
Since @modmuss50 pointed out in #674 that these proposed names should not exist in the tiny gzip'd data, I decided to decompress the tiny data, put enum names in, and then finally map the jar.

Signed-off-by: liach <liach@users.noreply.github.com>

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -408,7 +411,33 @@ task mapNamedJar(dependsOn: mapIntermediaryJar) {
 		logger.lifecycle(":mapping minecraft to named")
 
 		def tinyInput = new File("build/libs/yarn-tiny-${yarnVersion}.gz")
-		mapJar(namedJar, intermediaryJar, tinyInput, libraries, "intermediary", "named")
+
+		def decompressedTinyInput = new File(tempDir, "decompressed-named.tiny")
```
Comment (modmuss50) [inline build.gradle]: any reason this isnt its own task?

**Review by modmuss50 — CHANGES_REQUESTED** [review]
I think this should become its own task for neatness sake. 

I will also wait until after 1.14.1 to merge this in.

**Comment by asiekierka** [comment]
This should be a separate task - the named JAR mapping is for verifying the correctness of *Yarn mappings*, not *augmented Yarn mappings*.

File: README.md [hunk 2]
```diff
@@ -20,14 +20,28 @@ However, it is a good idea to consult name changes with other people - use pull
 ## Gradle
 Yarn uses Gradle to provide a number of utility tasks for working with the mappings.
 
+### `yarn`
+[`setupYarn`](#setupYarn) and download and launch the latest version of [Enigma](https://github.com/FabricMC/Enigma) automatically configured to use the merged jar and the mappings.
+
+Compared to launching Enigma externally, the gradle task adds a name guesser plugin that automatically map enums and a few constant field names.
```
Comment (modmuss50) [inline README.md:31]: > name guesser plugin

could explain it a bit better, is this info needed?

File: README.md [hunk 3]
```diff
@@ -20,14 +20,28 @@ However, it is a good idea to consult name changes with other people - use pull
...
 ### `setupYarn`
-`download` and `mergeJars`
+[`download`](#download) and [`mergeJars`](#mergeJars)
 
-### `yarn`
-`setupYarn` and download and launch the latest version of Enigma automatically configured to use the merged jar and the mappings.
+### `buildEnigma`
+Packages enigma mappings into a zip archive.
+
+### `compressTiny`
```
Comment (modmuss50) [inline README.md]: not sure this is needed, the whole idea is to just run gradle build

File: build.gradle (hunk 1 above)
Comment (liach) [reply build.gradle]: Separated now.

File: build.gradle [hunk 4]
```diff
@@ -397,21 +398,48 @@ task mapIntermediaryJar(dependsOn: [downloadMcLibs, build]) {
...
+	//Force the task to always run
+	outputs.upToDateWhen { false }
+
+	doLast {
+		logger.lifecycle(":mapping minecraft to named")
+
+		def resultMappings = mergeTiny.fileOutput
+		def namedWithEnum = new File(tempDir, "named-with-enum.tiny")
+
+		logger.lifecycle(":seeking auto-mappable fields")
```
Comment (modmuss50) [inline build.gradle:445]: I would make a separate task to generate this mapping file.

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Getting there, the documentation might need some cleanup, im not sure we need to document every task that 99% of people wont use. 

I would also make a new task that generates the mappings.

File: build.gradle (hunk 4 above)
Comment (liach) [reply build.gradle:445]: Didn't you reject the generation for its potential risk to infringe on mojang's copyright?

Comment (modmuss50) [reply build.gradle:445]: No need to ship it, just make it a task that generates it, its not the end of the world, I just think it makes more sense.

Comment (liach) [reply build.gradle:445]: Moved to another task.

File: build.gradle [hunk 5]
```diff
@@ -339,7 +357,6 @@ task compressTiny(dependsOn: ["tinyJar", "mergeTiny"], type: FileOutput){
 		fileInputStream.close()
 		outputStream.finish()
 		outputStream.close()
-		inputFile.delete()
```
Comment (modmuss50) [inline build.gradle:342]: Ill need to double check this wasnt important

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Gradle stuff looks good, need to test it. 

But I do not see any reason to have `buildEnigma` and `compressTiny` in the readme as `build` is already present.

I will test the gradle stuff myself in a bit before merging it in.

File: build.gradle (hunk 5 above)
Comment (liach) [reply build.gradle:342]: The task deleted this file; I retain the file so that secondary enum field filling could use it.

Comment (modmuss50) [reply build.gradle:342]: ah ok, thats fine.

File: README.md (hunk 2 above)
Comment (liach) [reply README.md:31]: Now i just use the latter description. Should suffice.

File: README.md (hunk 3 above)
Comment (liach) [reply README.md]: Build produces 2 artifacts, so I am explaining one by one

**Comment by liach** [comment]
Should I tweak this after FabricMC/Enigma#137 or FabricMC/Enigma#138 is merged so that we can use the new task? I will exclude the broken procyon from the dep.

**Comment by liach** [comment]
Note to @modmuss50: I have found a way to use new enigma features while holding the procyon dependency back. Will come back to this after FabricMC/Enigma#138 is merged so that we can use the command from enigma lib directly.

**Comment by liach** [comment]
@modmuss50 @gegy1000 Pr updated for the latest enigma. Uses latest enigma but still uses the good old procyon. Fixed package mapping errors. Please consider this one before others as this enhances our infrastructure and allows us to do better in the future.

**Comment by modmuss50** [comment]
Package fixes should be in another PR, and not mixed in with the gradle changes

File: README.md (hunk 3 above)
Comment (modmuss50) [reply README.md]: Doesnt matter, there should be no reason for anyone to run compressTiny, just use build

File: build.gradle [hunk 6]
```diff
@@ -50,40 +54,56 @@ configurations {
 
 dependencies {
 	enigmaRuntime "net.fabricmc:stitch:0.1.2.49+"
-	enigmaRuntime "cuchaz:enigma:0.13.1.99:all"
+
+	// not transitive
```
Comment (modmuss50) [inline build.gradle]: Why dont you just use the full version of engima? this seems horrible to maintain

File: build.gradle [hunk 7]
```diff
@@ -50,40 +54,56 @@ configurations {
...
+	enigmaRuntime name: "darcula", version: "1.0.0"
+	enigmaRuntime 'de.sciss:syntaxpane:1.2.+'
+	enigmaRuntime 'me.xdrop:fuzzywuzzy:1.2.0'
+	enigmaRuntime "org.bitbucket.mstrobel:procyon-compilertools:0.5.33.7"
+	//enigmaRuntime 'net.fabricmc:procyon-fabric-compilertools:0.5.33-SNAPSHOT'
+	// Uncomment when Rune fixed it
+	enigmaRuntime("cuchaz:enigma:0.13.1.114+") {
```
Comment (modmuss50) [inline build.gradle]: Here, you have the enigma version in 2 places, it should only be defined once

File: build.gradle [hunk 8]
```diff
@@ -206,33 +231,34 @@ task yarn(dependsOn: setupYarn) {
...
-		String[] args = [
-			mergedFile.getAbsolutePath(),
-			mappingsDir.getAbsolutePath()
-		]
+	outputs.upToDateWhen {false}
 
-		new CommandFindMappingErrors().run(args)
+	doLast {
+		logger.lifecycle(":checking mappings")
+		new CheckMappingsCommand().run mergedFile.getAbsolutePath(), mappingsDir.getAbsolutePath()
```
Comment (modmuss50) [inline build.gradle]: Put the brackets into this, I didnt know this was possible, but keep it consisitant with the rest of the file.

File: build.gradle [hunk 9]
```diff
@@ -339,19 +369,20 @@ task compressTiny(dependsOn: ["tinyJar", "mergeTiny"], type: FileOutput){
...
 		outputStream.close()
-		inputFile.delete()
 	}
 }
 
 clean.doFirst {
 	delete tempDir, cacheFilesMinecraft
 }
 
+tasks.build.dependsOn "checkMappings"
```
Comment (modmuss50) [inline build.gradle]: no, this should always be ran as a separate task and should not stop build from working.

File: mappings/net/minecraft/block/Block.mapping [hunk 10]
```diff
@@ -433,7 +433,7 @@ CLASS bmm net/minecraft/block/Block
 	METHOD h hasBlockEntity ()Z
```
Comment (modmuss50) [inline mappings/net/minecraft/block/Block.mapping]: Move all the mapping changes out of this PR please and into their own

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Main thing is to move the mapping changes out of this PR.

Im not too happy with how you have got enigma into the build-script as its going to be awful to maintain like this, use the all jar.

I still dont see the need to have some of the tasks that will never need to be ran by 99% of the people in the read me.

File: build.gradle (hunk 6 above)
Comment (liach) [reply build.gradle]: because full version ships the wrong procyon. Please!

File: build.gradle (hunk 8 above)
Comment (liach) [reply build.gradle]: :sweat:

File: build.gradle (hunk 7 above)
Comment (liach) [reply build.gradle]: Is it even possible to use project properties in buildscript block? I need to see

File: mappings/net/minecraft/block/Block.mapping (hunk 10 above)
Comment (liach) [reply mappings/net/minecraft/block/Block.mapping]: Removed.

File: build.gradle (hunk 6 above)
Comment (modmuss50) [reply build.gradle]: Fix engima then ;)

File: build.gradle (hunk 9 above)
Comment (liach) [reply build.gradle]: Fixed locally.

File: build.gradle [hunk 11]
```diff
@@ -30,13 +31,16 @@ def build_number = ENV.BUILD_NUMBER ?: "local"
...
 def yarnVersion = "${minecraft_version}+build.$build_number"
 
-
 repositories {
 	mavenCentral()
 	maven {
 		name "Modmuss50 Repository"
 		url 'https://maven.modmuss50.me'
 	}
+	ivy {
```
Comment (modmuss50) [inline build.gradle]: I dont think this is needed now

File: build.gradle [hunk 12]
```diff
@@ -50,40 +54,46 @@ configurations {
 
 dependencies {
 	enigmaRuntime "net.fabricmc:stitch:0.1.2.49+"
-	enigmaRuntime "cuchaz:enigma:0.13.1.99:all"
+	enigmaRuntime "cuchaz:enigma:0.13.1.114+:all" // Update version above as well!
```
Comment (modmuss50) [inline build.gradle]: is 114+ correct? should it be `0.13.1.+`?

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Getting close, branch also needs updating.

File: build.gradle (hunk 12 above)
Comment (liach) [reply build.gradle]: Still locked on enigma. @modmuss50 Can you do us a favor by reverting https://github.com/FabricMC/Enigma/commit/44dd51db443839788fe23c55d6036d8fa3ba692b ?

**Comment by liach** [comment]
@modmuss50 fixed everything and limited scope. Please review this zombie pr again.

**Comment by modmuss50** [comment]
Looks good, are we still waiting on fixing Procyon? or reverting it?

**Comment by liach** [comment]
procyon is still broken for now, but it is not in the scope of this pr any more. i need to learn procyon like runemoro did in order to fix that.

File: build.gradle [hunk 13]
```diff
@@ -206,33 +215,22 @@ task yarn(dependsOn: setupYarn) {
...
 
 task buildEnigma(type: Zip) {
+	group = buildMappingGroup
 	from mappingsDir
 	include "**/*"
 	archiveName "yarn-enigma-${yarnVersion}.zip"
 	destinationDir(file("build/libs"))
 }
 
-task checkMappings {
```
Comment (modmuss50) [inline build.gradle]: Add this back for now, it can be looked at more in another PR.

File: build.gradle [hunk 14]
```diff
@@ -397,21 +400,61 @@ task mapIntermediaryJar(dependsOn: [downloadMcLibs, build]) {
...
+	inputs.files downloadMcLibs.outputs.files.files
+	outputs.file(namedJar)
+
+	//Force the task to always run
+	outputs.upToDateWhen { false }
+
+	doLast {
+		logger.lifecycle(":mapping minecraft to named")
+
+		def namedWithEnum = buildTinyWithEnum.fileOutput
```
Comment (modmuss50) [inline build.gradle]: Dont really need this as its own def, not a big thing.

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Very close to me approving it, take a look at theses things.

Can anyone else take a look seems easy enough. The group thing doesnt seem to provide a lot, but I dont see any harm in having it.

File: build.gradle (hunk 13 above)
Comment (liach) [reply build.gradle]: Fixed.

File: build.gradle (hunk 14 above)
Comment (liach) [reply build.gradle]: Fixed.

**Review by modmuss50 — APPROVED** [review]
Yep, im happy with this, would be good to get some feedback form someone else as well.

File: build.gradle [hunk 15]
```diff
@@ -227,12 +239,14 @@ task checkMappings {
 }
 
 task downloadIntermediary(type: Download){
-    def url = "https://github.com/FabricMC/intermediary/raw/master/mappings/${minecraft_version}.tiny"
-    src url.replaceAll(" ", "%20")
-    dest new File(cacheFilesMinecraft, "${minecraft_version}-intermediary.tiny")
+	group = buildMappingGroup
+	def url = "https://github.com/FabricMC/intermediary/raw/master/mappings/${minecraft_version}.tiny"
+	src url.replaceAll(" ", "%20")
```
Comment (ChloeDawn) [inline build.gradle]:
```suggestion
	src com.google.common.net.UrlEscapers.urlFragmentEscaper().escape(url)
```
This should be replaced with [Guava's UrlEscapers API](https://google.github.io/guava/releases/23.0/api/docs/com/google/common/net/UrlEscapers.html).

**Review by ChloeDawn — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 15 above)
Comment (liach) [reply build.gradle]: Resolved. We use imports in build gradle btw.

Comment (ChloeDawn) [reply build.gradle]: I included the package so you knew what to import, and I can't suggest changes for multiple arbitrary lines with the suggestion system 😜

Comment (liach) [reply build.gradle]: Ah, github was dumb as usual :trollface:

**Comment by Runemoro** [comment]
> This should be a separate task - the named JAR mapping is for verifying the correctness of Yarn mappings, not augmented Yarn mappings.

I'd say that if a manually mapped name conflicts with an "augmented" name, the mappings are broken. We've had problems in the past due to a field being accidentally named the same as an enum constant. The checkMappings task should check this.

File: build.gradle (hunk 15 above)
Comment (modmuss50) [reply build.gradle]: Good idea, this was done in a rush ;)

**Comment by modmuss50** [comment]
Merged, not sure why its not shown. Thanks.

**Comment by liach** [comment]
Thanks :heart: It just broke for the base branch is pre2 instead of pre4 (I should have updated that)
